tianshou/data/buffer.py

```diff
@@ -241,117 +191,111 @@ def add(
         info: Optional[Union[dict, Batch]] = {},
         policy: Optional[Union[dict, Batch]] = {},
         **kwargs: Any,
-    ) -> None:
-        """Add a batch of data into replay buffer."""
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """Add a batch of data into replay buffer.
+
+        Return (episode_length, episode_reward) if one episode is terminated,
+        otherwise return (0, 0.0).
+        """
         assert isinstance(
             info, (dict, Batch)
         ), "You should return a dict in the last argument of env.step()."
-        if self._last_obs:
+        if self._save_only_last_obs:
             obs = obs[-1]
         self._add_to_buffer("obs", obs)
         self._add_to_buffer("act", act)
-        # make sure the reward is a float instead of an int
-        self._add_to_buffer("rew", rew * 1.0)  # type: ignore
-        self._add_to_buffer("done", done)
-        if self._save_s_:
+        # make sure the data type of reward is float instead of int
+        rew = rew * 1.0  # type: ignore
+        self._add_to_buffer("rew", rew)
```

Comment:
better just do `self._add_to_buffer("rew", float(rew))` and remove all this mess (including comment since it is self-explainatory)

Reply:
`rew` may be `np.ndarray`:
```
In [1]: import numpy as np                                                                                                                                                                                          

In [2]: a=np.array([0,0,0])                                                                                                                                                                                         

In [3]: float(a)                                                                                                                                                                                                    
`---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-6c8cade3f87c> in <module>
----> 1 float(a)

TypeError: only size-1 arrays can be converted to Python scalars
``

Reply:
AFAIK, the reward must be a scalar according to OpenAI gym API.

Reply:
Indeed, but we currently have a unittest for multi-agent (multi-agent env's reward is np.ndarray)